tests/puppeteer/pages/BO/product.js

```diff
@@ -1,13 +1,116 @@
+// Using chai
+const {expect} = require('chai');
+// Importing page
 const BOBasePage = require('../BO/BObasePage');
 
 module.exports = class Product extends BOBasePage {
   constructor(page) {
     super(page);
 
     this.pageTitle = 'Products •';
+    this.productDeletedSuccessfulMessage = 'Product successfully deleted.';
+
+    // Selectors
+    // List of products
+    this.productListForm = '#product_catalog_list';
+    this.productRow = `${this.productListForm} table tbody tr`;
+    // Filters input
+    this.productFilterInput = `${this.productListForm} input[name='filter_column_%FilterBy']`;
+    this.filterSearchButton = `${this.productListForm} button[name='products_filter_submit']`;
+    this.filterResetButton = `${this.productListForm} button[name='products_filter_reset']`;
+    // Filter Category
+    this.treeCategoriesBloc = `#tree-categories`;
+    this.filterByCategoriesButton = `#product_catalog_category_tree_filter button`;
+    this.filterByCategoriesExpandButton = `${this.treeCategoriesBloc} a#product_catalog_category_tree_filter_expand`;
+    this.filterByCategoriesUnselectButton = `${this.treeCategoriesBloc} a#product_catalog_category_tree_filter_reset`;
+    this.filterByCategoriesCategoryLabel = `${this.treeCategoriesBloc} label.category-label`;
+    // HEADER buttons
+    this.addProductButton = `#page-header-desc-configuration-add`;
   }
 
   /*
   Methods
    */
+
+  /**
+   * Filter products from inputs : Name, reference and Category
+   * @param filterBy
+   * @param value
+   * @return {Promise<void>}
+   */
+  async filterProducts(filterBy, value = '') {
+    await this.page.type(this.productFilterInput.replace('%FilterBy', filterBy), value);
+    await Promise.all([
+      this.page.waitForNavigation({waitUntil: 'networkidle0'}),
+      this.page.click(this.filterSearchButton),
+    ]);
+  }
+
+  /**
+   * Get number of products displayed in list
+   * @return integer
+   */
+  async getNumberOfProductsFromList() {
+    return (await this.page.$$(this.productRow)).length;
```

Comment:
Isn't this limited by pagination ? Maybe we can get the total number from the page itself (IIRC it's written somewhere) ?

Reply:
we have 19 demo products and 20 items/page showed so if we destroyed each product created, we will not have to navigate between pages

Reply:
At the moment, yes. But I prefer we not rely on the fixtures as there are today because we don't know how they might change. If the dev team modify the fixtures to add 25 demo products, this will fail.

Reply:
The only way I see to do it is this 
```js
    async getNumberOfProductsFromList() {
    const numberOfProducts = (await this.page.$$(this.productRow)).length;
    const found = await this.elementVisible(this.paginationNextLink, 1000);
    if (!found) return numberOfProducts;

    await Promise.all([
      this.page.waitForNavigation({waitUntil: 'networkidle0'}),
      this.page.click(this.paginationNextLink),
    ]);
    return numberOfProducts + (await this.getNumberOfProductsFromList());
  }
```

Reply:
The thing is, if there's more than 20 articles, you'll see this block at the bottom : 
![image](https://user-images.githubusercontent.com/49909275/64117431-7d72af80-cd95-11e9-8523-8dc3492de0b6.png)
It gives you the number of items. Maybe we can try looking for it to see if there's more than 20 items ?

Reply:
that's why I click paginationNextLink (if exists).
This recursive function will return the real number of products from all products.

Reply:
Your method will crawl every page, and might take a while if there's a lot of pages. 
- if there's not a pagination block, we count the number of rows
- if there's a pagination block, we extract the number of lines from the text inside with a regexp (like `/out of ([0-9]+)/` )
Both methods don't require any navigation.

What do you think ?

Reply:
It's better to not have navigation for sure.
It would be better if we didn't have to do regex to get it.
I changed it like suggested with regex, and looking forward to get it with a better selector.